tests/python/pants_test/engine/test_scheduler.py

```diff
@@ -57,6 +60,65 @@ def transitive_coroutine_rule(c):
   yield D(b)
 
 
+@union
```

Comment:
Doing this part first after rebasing!

Reply:
Did this and it seemed to mysteriously work? It essentially just let us remove the `if` in this comprehension you've highlighted.

Reply:
Looks great!

One possibility (not sure whether it is an improvement though...) would be to do a function call rather than a decorator for union members.

That might look like:
```
@union
class HydrateableField(object): pass

class SourcesField(...): ...

def rules():
  return [
      union_rule(HydrateableField, SourcesField)
    ]
```

Reply:
There ended up being a ridiculous amount of confusion when trying to support use of `@union_rule` as a decorator and as a function call, so I instead converted it to a datatype `UnionRule` like we do with e.g. `RootRule`, without any decorators, which makes much more sense anyway. I just didn't want to drop the possibility of it being a decorator too early.